src/main/java/io/supertokens/session/accessToken/AccessTokenSigningKey.java

```diff
@@ -69,142 +73,206 @@ public static AccessTokenSigningKey getInstance(Main main) {
+                        if (config.getAccessTokenSigningKeyDynamic()) {
+                            lastCreated = lastCreated < key.createdAtTime ? key.createdAtTime : lastCreated;
+                            
+                            if (keyCurrentAfter <= key.createdAtTime) {
+                                generateNewKey = false;
+                            } 
+                            if (key.createdAtTime < keysValidAfter) {
+                                clearKeys = true;
+                            } else {
+                                validKeys.add(new KeyInfo(key.value, key.createdAtTime, validityDuration));
+                            }
+                        }
                     }
                 }
 
-                if (key == null || generateNewKey) {
+                if (generateNewKey) {
                     String signingKey;
                     try {
                         Utils.PubPriKey rsaKeys = Utils.generateNewPubPriKey();
                         signingKey = rsaKeys.toString();
                     } catch (NoSuchAlgorithmException e) {
                         throw new StorageTransactionLogicException(e);
                     }
-                    key = new KeyInfo(signingKey, System.currentTimeMillis());
-                    boolean success = noSQLStorage.setAccessTokenSigningKey_Transaction(
-                            new KeyValueInfoWithLastUpdated(key.value, key.createdAtTime,
-                                    keyFromStorage == null ? null : keyFromStorage.lastUpdatedSign));
+                    KeyInfo newKey = new KeyInfo(signingKey, System.currentTimeMillis(), validityDuration);
+                    boolean success = noSQLStorage.addAccessTokenSigningKey_Transaction(
+                        new KeyValueInfo(newKey.value, newKey.createdAtTime), lastCreated
+                    );
+                    validKeys.add(newKey);
+
                     if (!success) {
                         // something else already updated this particular field. So we must try again.
                         continue;
                     }
                 }
-
-                return key;
-
+                
+                if (clearKeys) {
+                    noSQLStorage.removeAccessTokenSigningKeysBefore_Transaction(keysValidAfter);
+                }
             }
+        } else {
+            throw new QuitProgramException("Unsupported storage type detected");
         }
 
-        throw new QuitProgramException("Unsupported storage type detected");
+        validKeys.sort(Comparator.comparingLong((KeyInfo key) -> key.createdAtTime).reversed());
+
+        return Collections.unmodifiableList(validKeys);
     }
 
     public static class KeyInfo {
         public String value;
         public long createdAtTime;
+        public long expiryTime;
 
-        KeyInfo(String value, long createdAtTime) {
+        KeyInfo(String value, long createdAtTime, long validityDuration) {
```

Comment:
Why do we need to store `validityDuration` in this object?

Reply:
It isn't strictly necessary, but we need this info when constructing API responses it's already calculated here.